runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java

```diff
@@ -286,9 +317,55 @@ public void process(ApexStreamTuple<WindowedValue<Iterable<?>>> t) {
     }
   }
 
+  @Override
+  public void fireTimer(Object key, Collection<TimerData> timerDataSet) {
+    pushbackDoFnRunner.startBundle();
+    @SuppressWarnings("unchecked")
+    Coder<Object> keyCoder = (Coder) currentKeyStateInternals.getKeyCoder();
+    ((StateInternalsProxy) currentKeyStateInternals).setKey(key);
+    currentKeyTimerInternals.setContext(key, keyCoder, new Instant(this.currentInputWatermark),
+        new Instant(this.currentOutputWatermark));
+    for (TimerData timerData : timerDataSet) {
+
+      if (timerData.getDomain() == TimeDomain.PROCESSING_TIME) {
+        System.out.println(
+            Thread.currentThread().getName() + "\n\n+++fireTimer: key=" + key + " " + timerData);
+        System.out.println("current time is: "
+            + new java.util.Date(currentKeyTimerInternals.currentProcessingTime().getMillis()));
+      }
+
+      StateNamespace namespace = timerData.getNamespace();
+      checkArgument(namespace instanceof WindowNamespace);
+      BoundedWindow window = ((WindowNamespace<?>) namespace).getWindow();
+      pushbackDoFnRunner.onTimer(timerData.getTimerId(), window,
+          timerData.getTimestamp(), timerData.getDomain());
+    }
+    pushbackDoFnRunner.finishBundle();
+  }
+
   private void processWatermark(ApexStreamTuple.WatermarkTuple<?> mark) {
     this.currentInputWatermark = mark.getTimestamp();
-
+    long minEventTimeTimer = currentKeyTimerInternals.fireReadyTimers(
+        this.currentInputWatermark,
+        this, TimeDomain.EVENT_TIME);
+
+    checkState(minEventTimeTimer >= currentInputWatermark,
+        "Event time timer processing generates new timer(s) behind watermark.");
+    //LOG.info("Processing time timer {} registered behind watermark {}", minProcessingTimeTimer,
+    //    currentInputWatermark);
+
+    // TODO: is this the right way to trigger processing time timers?
+    // drain all timers below current watermark, including those that result from firing
+    long minProcessingTimeTimer = Long.MIN_VALUE;
+    while (minProcessingTimeTimer < currentInputWatermark) {
```

Comment:
This doesn't seem like a correct way of firing processing time timers:

- Their timestamps should be compared to Instant.now() rather than to the current input watermark, because the watermark is in event time domain.
- They should be fired probably somewhere outside processWatermark(), because they are triggered by real-world time passing, rather than changes of watermark - again because of the time domain.